osbuild/ostree.py

```diff
@@ -0,0 +1,90 @@
+import contextlib
+import json
+import os
+import tempfile
+
+from typing import List
+
+
+class Param:
+    def __init__(self, value_type, mandatory=False, default=None):
+        self.type = value_type
+        self.mandatory = mandatory
+        self.default = default
+
+    def check(self, value):
+        origin = getattr(self.type, "__origin__", None)
+        if origin:
+            self.typecheck(value, origin)
+            if origin is list:
+                self.check_list(value, self.type)
+            else:
+                raise NotImplementedError(origin)
+        else:
+            self.typecheck(value, self.type)
+
+    @staticmethod
+    def check_list(value, tp):
+        inner = tp.__args__
+        for x in value:
+            Param.typecheck(x, inner)
+
+    @staticmethod
+    def typecheck(value, tp):
+        if isinstance(value, tp):
+            return
+        raise ValueError(f"{value} is not of {tp}")
+
+
+class Treefile:
+    parameters = {
+        "ref": Param(str, True),
+        "gpg-key": Param(str),
+        "repos": Param(List[str], True),
+        "selinux": Param(bool),
+        "boot-location": Param(str),
+        "etc-group-members": Param(List[str]),
+        "ignore-removed-users": Param(List[str]),
+        "ignore-removed-groups": Param(List[str]),
+        "tmp-is-dir": Param(bool),
+        "machineid-compat": Param(bool)
+    }
+
+    def __init__(self):
+        self._data = {}
+        self["ref"] = "osbuild/devel"
+        self["repos"] = ["osbuild"]
+
+    def __getitem__(self, key):
+        param = self.parameters.get(key)
+        if not param:
+            raise ValueError(f"Unknown param: {key}")
+        return self._data[key]
+
+    def __setitem__(self, key, value):
+        param = self.parameters.get(key)
+        if not param:
+            raise ValueError(f"Unknown param: {key}")
+        param.check(value)
+        self._data[key] = value
+
+    def dumps(self):
+        return json.dumps(self._data)
+
+    def dump(self, fp):
+        return json.dump(self._data, fp)
+
+    @contextlib.contextmanager
+    def as_tmp_file(self):
+        name = None
+        try:
+            fd, name = tempfile.mkstemp(suffix=".json",
+                                        text=True)
+
+            with os.fdopen(fd, "w+") as f:
```

Comment:
Does `fdopen()` close the fd even if it raises an exception for other reasons? The docs are not really clear, and I got lost in the c-sources. But we also probably don't really care, as we re-raise the exception anyway.

Reply:
Good that you asked, I was wondering exactly the same when I wrote the code but then forgot about it again. Sadly that means I also don't know ;)

Reply:
Hehe, I see. I checked the sources and you can easily make `io.open()` trigger an exception without closing the passed file-descriptor [1]. But you can also cause it to close it [2]. But I think as long as the arguments are valid, it will always end up closing it under `OSError`s. Regardless, nothing we can do about it, other than assuming it cleaned it up correctly.

[1] https://github.com/python/cpython/blob/master/Lib/_pyio.py#L170
[2] https://github.com/python/cpython/blob/master/Lib/_pyio.py#L256